openedx/openedx-platform
# MIT CAPA improvements: Add show_correctness field

Adds `Correctness Visibility` setting to subsections, which lets course authors control when problem correctness, score, adaptive hints, and answers are shown to learners.

This setting was previously available in problem-level XML for Multiple Choice and Checkbox problems (cf [DOC-1319](https://openedx.atlassian.net/browse/DOC-1319), [DOC-832](https://openedx.atlassian.net/browse/DOC-832)), but correctness still leaked through on the problem page in the score area, notifications area, answers, and in AJAX response data.

This change replaces and improves the problem-level XML setting, but does not disrupt the legacy behaviour.

**JIRA tickets**:

* [OSPR-1716](https://openedx.atlassian.net/browse/OSPR-1716)

**Discussions**:

[Technical discovery document](https://docs.google.com/document/d/1yGAbIl5lZUXB_BTFDp5dx75fEypQHA7Ipg7M9PATKaw/edit)

**Dependencies**:

None.

Related: https://github.com/edx/edx-documentation/pull/1466

**Screenshots**:

Studio subsection modal for Instructor-paced courses, using new Visibility tab:
![screen shot 2017-04-13 at 9 00 45 am](https://cloud.githubusercontent.com/assets/7556571/24983662/c17cea92-2027-11e7-86fa-cba9a028cecb.png)

Studio subsection modal for self-paced courses (note different text for Past Due, and Special Exams not enabled, so no Advanced tab):
![screen shot 2017-04-13 at 9 01 39 am](https://cloud.githubusercontent.com/assets/7556571/24983680/ea400324-2027-11e7-9e06-a5cc4a68ff77.png)

Problem score, correctness, and answer is hidden from GUI in "progress" and "notification" areas of the problem pages:

cf https://github.com/open-craft/edx-platform/pull/58, https://github.com/open-craft/edx-platform/pull/59

![screen shot 2017-04-13 at 9 02 45 am](https://cloud.githubusercontent.com/assets/7556571/24983699/0bce70d4-2028-11e7-9784-d9a2e5a7a5b0.png)

Problem score and correctness hidden from AJAX:

cf https://github.com/open-craft/edx-platform/pull/61

**Sandbox URLs**:

* Studio: https://studio-pr14737.sandbox.opencraft.hosting/
* LMS: https://pr14737.sandbox.opencraft.hosting/

Latest commit: 6e8a235b0330cc3677e15d84d5496da88a052734

* Instructor-paced [MIT CAPA1 test course](https://pr14737.sandbox.opencraft.hosting/courses/course-v1:OpenCraft+CAPA1+2017/courseware) created with [show_correctness.course.tar.gz](https://github.com/edx/edx-platform/files/915805/show_correctness.course.tar.gz)
* Self-paced [MIT CAPA2 test course](https://pr14737.sandbox.opencraft.hosting/courses/course-v1:OpenCraft+CAPA2+selfpace/courseware) created with [self_paced.show_correctness.course.tar.gz](https://github.com/edx/edx-platform/files/915806/self_paced.show_correctness.course.tar.gz)

**Partner information**:

3rd party-hosted open edX instance

**Deployment targets**:

* edx.org
* edge.edx.org

**Merge deadline**:

April 21st, 2017

**Testing instructions**:

* Results Visibility setting on Subsection: see https://github.com/open-craft/edx-platform/pull/60
* Problem score and correctness is withheld in "score progress" section: see https://github.com/open-craft/edx-platform/pull/58
* Notification text withholds correctness: see https://github.com/open-craft/edx-platform/pull/59
* AJAX withholds correctness: see https://github.com/open-craft/edx-platform/pull/61

**Author concerns and questions**

1. Course Progress page has not been updated; will be addressed in a subsequent PR.
1. Do we need to add a visual indicator in Studio of the "show correctness" setting change on the subsection itself? E.g., when we change Subsection Visibility, we can see a message like "Subsection is hidden after due date". If we need this change, then we may also need an icon to show with it.

**Reviewers**:

- [x] OpenCraft: @e-kolpakov, @itsjeyd  
- [x] UX: @srpearce  
- [ ] Eng: @cahrens 

CC @jbarciauskas , @bradenmacdonald , @edx/edx-accessibility, @srpearce
